transport/http2_client.go

```diff
@@ -558,7 +572,18 @@ func (t *http2Client) handleSettings(f *http2.SettingsFrame) {
 			defer t.mu.Unlock()
 			switch s.ID {
 			case http2.SettingMaxConcurrentStreams:
-				t.maxStreams = v
+				// TODO(zhaoq): This is a hack to avoid significant refactoring of the
+				// code to deal with the unrealistic int32 overflow. Probably will try
+				// to find a better way to handle this later.
+				if v > math.MaxInt32 {
+					v = math.MaxInt32
+				}
+				if t.streamsQuota == nil {
+					t.streamsQuota = newQuotaPool(int(v))
+				} else {
+					t.streamsQuota.reset(int(v) - t.maxStreams)
```

Comment:
I do not find issue here. For example, the new quota is 10 (v) and the old quota is 100 (t.maxStream). And line 255 wants to return 99 back to stream quota (100 -1). If the reset occurs before the add, we get t.streamsQuota == -90 after the reset and 9 quota on the t.streamsQuota.acquire() and t.streamsQuota == 9, which are all expected. Note that v-t.maxStreams is not the configured amount. Instead v is -- v - t.maxStreams + x < v because 0 < x <= t.maxStreams -1.